adapters/ads_adapter.py

```diff
@@ -1,22 +1,173 @@
+        self.notification_handles = []
+        self.plc = pyads.Connection(ams_net_id, ams_net_port)
+        self.plc.open()
+
+    def __del__(self):
+        for handle in self.notification_handles:
+            self.plc.del_device_notification(*handle)
+
+        self.plc.close()
+
+    def log(self, name: str, plc_type):
+        attr = pyads.NotificationAttrib(ctypes.sizeof(plc_type))
+        handles = self.plc.add_device_notification(
+            name, attr, self.callback(plc_type), pyads.ADSTRANS_SERVERCYCLE
+        )
+        self.notification_handles.append(handles)
+
+    def subscribe(self, pattern: str):
+        parsed_entries = self.get_ads_entries()
+
+        for entry in parsed_entries:
+            if re.match(pattern, entry.name) != None:
+                self.log(entry.name, self.DATATYPE_MAP[entry.typename])
+
+    def callback(self, plc_type):
+        @self.plc.notification(plc_type)
+        def decorated_callback(handle, name, timestamp, value):
+            client.send_sample(
+                name, AdsClient.format_timestamp(timestamp), float(value),
+            )
+
+        return decorated_callback
+
+    def get_ads_entries(self):
+        upload_info = self._get_upload_info()
+        entries = self.plc.read(
+            pyads.constants.ADSIGRP_SYM_UPLOAD, 0, ctypes.c_ubyte * upload_info.nSymSize
+        )
+
+        parsed_entries = self._unpack_entries(
+            bytes(entries), upload_info.nSymSize, upload_info.nSymbols
+        )
+        return parsed_entries
+
+    @staticmethod
+    def format_timestamp(timestamp: datetime):
+        return time.mktime(timestamp.timetuple()) + timestamp.microsecond / 1e6
+
+    def _get_upload_info(self):
+        upload_info = self.plc.read(
+            pyads.constants.ADSIGRP_SYM_UPLOADINFO,
+            0,
+            pyads.structs.SAdsSymbolUploadInfo,
+        )
+        return upload_info
+
+    def _unpack_entry(data: bytes, index: int) -> (int, Entry):
+        fmt = "<6I3H"
+        (
+            entry_length,
+            _,
+            _,
+            entry_datatype_size,
+            entry_datatype,
+            _,
+            name_len,
+            type_len,
+            comment_len,
+        ) = struct.unpack_from(fmt, entries_data, index)
+
+        name_offset = index + struct.calcsize(fmt)
+        name = entries_data[name_offset : name_offset + name_len].decode("ascii")
+
+        type_offset = name_offset + name_len + 1
+        typ = entries_data[type_offset : type_offset + type_len].decode("ascii")
+
+        comment_offset = type_offset + type_len + 1
+        comment = entries_data[comment_offset : comment_offset + comment_len].decode(
+            "ascii"
+        )
+
+        return (
+            entry_length,
+            self.Entry(name, typ, comment, entry_datatype, entry_datatype_size),
+        )
+
+    def _unpack_entries(self, entries_data: bytes, size: int, count: int):
```

Comment:
Since this function is unrelated to this class, it might make sense to move it out of the class and make it a simple function.